Add support for users updating the CSP's sources

Sigmonia opened the pull request:
#### Rationale
[Secure Issue 50544: UI-based support for customizing CSP-allowed servers](https://www.labkey.org/home/Developer/issues/Secure/issues-details.view?issueId=50544)
I added a new link in the Admin Console > Settings > Configuration column and reused the External Redirect UI, forms, and controller. 

#### Related PRs
* https://github.com/LabKey/testAutomation/pull/2105

#### Changes
- Generalized the ExternalRedirectAdminAction's JSPs, form, and methods
- Added a method to register the new source hosts with the CSP at startup and save

Sigmonia committed:
Add support for users updating the CSP's sources
- Add new admin link
- Update ExternalRedirectHosts to be more general
- Update Form and JSPs
- Register Sources on CoreModule load

Sigmonia committed:
rename vars to match form

labkey-adam commented on api/src/org/labkey/api/settings/AppPropsImpl.java:
```diff
@@ -625,7 +625,19 @@ public List<String> getExternalRedirectHosts()
+            return new ArrayList<>(Arrays.asList(urls.split(EXTERNAL_HOST_DELIMITER)));
         }
         return new ArrayList<>();
     }
```
Could factor out a common method that takes the enum constant to eliminate duplicate code

labkey-adam commented on api/src/org/labkey/api/settings/WriteableAppProps.java:
```diff
@@ -221,8 +221,18 @@ public void setExternalRedirectHosts(@NotNull List<String> externalRedirectHosts
+        }
+        storeStringValue(externalSourceHostURLs, urls.toString());
+    }
 }
```
Factor out a common method here as well?

labkey-adam commented on core/src/org/labkey/core/CoreModule.java:
```diff
@@ -1249,6 +1250,15 @@ public void moduleStartupComplete(ServletContext servletContext)
+                AdminController.getExternalSourceHostsKey(),
+                AppProps.getInstance().getExternalSourceHosts().toArray(new String[0]));
+
+        LOG.debug("Registered [{}] as an allowed connection source", String.join(", ", AppProps.getInstance().getExternalSourceHosts()));
```
```suggestion
        LOG.debug("Registered [{}] as an allowed connection source", () -> String.join(", ", AppProps.getInstance().getExternalSourceHosts()));

```

labkey-adam replied:
Totally optional, but this avoids doing the join work unless the level is `DEBUG`.

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
+        public ModelAndView getView(ExternalHostsForm form, boolean reshow, BindException errors)
         {
-            form.setExistingRedirectHostList(AppProps.getInstance().getExternalRedirectHosts());
+            form.setExistingHostsList(form.getTypeEnum() == ExternalServerType.Source ? AppProps.getInstance().getExternalSourceHosts() : AppProps.getInstance().getExternalRedirectHosts());
```
Cleaner (and more future expandable) would be to push the enum-specific details into each constant. So caller would be something like:
```
form.setExistingHostsList(form.getTypeEnum().getHostsList());
```

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
+            if (form.getTypeEnum() == ExternalServerType.Source)
+                update = handleHostPost(form, appProps::getExternalSourceHosts, appProps::setExternalSourceHosts, errors);
+            else
+                update = handleHostPost(form, appProps::getExternalRedirectHosts, appProps::setExternalRedirectHosts,  errors);
```
Ideally, this code wouldn't know anything about the specific behaviors of the enum constants... just pass the enum constant into `handleHostPost()` and implement `getHosts()` and `setHosts()` on the enum.

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
+                    // Refresh the CSP with new values.
+                    ContentSecurityPolicyFilter.unregisterAllowedConnectionSource(getExternalSourceHostsKey());
+                    ContentSecurityPolicyFilter.registerAllowedConnectionSource(getExternalSourceHostsKey(), appProps.getExternalSourceHosts().toArray(new String[0]));
+                }
```
Similarly, implement `afterSave()` on the enum and call it unconditionally. Empty method in the case of Redirect.

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
                 String urlToDelete = form.getExistingExternalHost();
-                List<String> redirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                List<String> redirectHosts = getHosts.get();
                 for (String externalRedirectHost : redirectHosts)
```
```suggestion
                for (String externalHost : externalHosts)

```

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
             {
                 String urlToDelete = form.getExistingExternalHost();
-                List<String> redirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                List<String> redirectHosts = getHosts.get();
```
```suggestion
                List<String> externalHosts = getHosts.get();

```

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
-                        WriteableAppProps appProps = AppProps.getWriteableInstance();
-                        appProps.setExternalRedirectHosts(redirectHosts);
-                        appProps.save(getUser());
+                        setHosts.accept(redirectHosts);
```
```suggestion
                        setHosts.accept(externalHosts);

```

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
             {
-                List<String> redirectHosts = form.getExistingRedirectHostList(); //get hosts from the form, this includes updated hosts
-                if (null != redirectHosts && redirectHosts.size() > 0)
+                List<String> redirectHosts = form.getExistingHostList(); //get hosts from the form, this includes updated hosts
```
Rename... `externalHosts`?

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10699,13 +10727,11 @@ else if (form.isSaveNew())
                 else if (StringUtils.isNotEmpty(newExternalRedirectHost))
                 {
-                    List<String> existingRedirectHosts = AppProps.getInstance().getExternalRedirectHosts();
+                    List<String> existingRedirectHosts = getHosts.get();
```
Rename?

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10690,7 +10718,7 @@ else if (form.isSaveAll())
             else if (form.isSaveNew())
             {
-                String newExternalRedirectHost = StringUtils.trimToEmpty(form.getNewExternalRedirectHost());
+                String newExternalRedirectHost = StringUtils.trimToEmpty(form.getNewExternalHost());
```
Rename

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
 
-            JspView<ExternalRedirectForm> newView = new JspView<>("/org/labkey/core/admin/addNewExternalRedirectHost.jsp", form, errors);
-            newView.setTitle("Register New External Redirect Host");
+            JspView<ExternalHostsForm> newView = new JspView<>("/org/labkey/core/admin/addNewExternalRedirectHost.jsp", form, errors);
```
Should probably change the JSP names

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10629,59 +10637,79 @@ private static class IgnorePasswordMixIn
             newView.setFrame(WebPartView.FrameType.PORTAL);
-            JspView<ExternalRedirectForm> existingView = new JspView<>("/org/labkey/core/admin/existingExternalRedirectHosts.jsp", form, errors);
-            existingView.setTitle("Existing External Redirect Hosts");
+            JspView<ExternalHostsForm> existingView = new JspView<>("/org/labkey/core/admin/existingExternalRedirectHosts.jsp", form, errors);
```
Should probably change the JSP names

labkey-adam commented on core/src/org/labkey/core/admin/AdminController.java:
```diff
@@ -10747,9 +10773,9 @@ private boolean isDuplicate(List<String> existingURLs, String newExternalRedirec
+            return new ActionURL(ExternalRedirectAdminAction.class, getContainer()).addParameter("type", form.getTypeEnum().name());
         }
 
         @Override
```
Need to update `addNavTrail()` as well. A help topic needs to be defined on the enum and used here. And the title needs to change based on the enum.

Sigmonia committed:
Addressing CR feedback
- DRY'd Action by moving most methods to Enum
- renamed JSPs, Action, and local vars containing "redirect"
- updated NavTrail and getSuccessURL methods to respect type parameter
- Added URL/Host validation

labkey-adam commented on api/src/org/labkey/api/settings/WriteableAppProps.java:
```diff
@@ -216,13 +216,18 @@ public void setXFrameOption(String option)
+
+    private void setExternalHosts(RandomStartupProperties propName, @NotNull Collection<String> externalSourceHosts)
+    {
+        storeStringValue(propName, String.join(EXTERNAL_HOST_DELIMITER, externalSourceHosts));
```
Previous code did a `trim()` on each url... no longer needed?

Sigmonia replied:
I don't believe so. I moved the trim to the validation & set-addition step. I guess that opens a potential hole if we add new callers, but I think "trimming" fits better as part of the validation/sanitation process.

labkey-adam replied:
Cool, I missed that

labkey-adam reviewed (approved):
One question about trimming hosts before storing

Sigmonia committed:
Adjust the help link topic
